arwen/contexts/output.go

```diff
@@ -285,21 +298,68 @@ func (context *outputContext) AddTxValueToAccount(address []byte, value *big.Int
 
 // GetVMOutput updates the current VMOutput and returns it
 func (context *outputContext) GetVMOutput() *vmcommon.VMOutput {
-	if context.outputState.ReturnCode == vmcommon.Ok {
-		runtime := context.host.Runtime()
-		metering := context.host.Metering()
+	runtime := context.host.Runtime()
+	metering := context.host.Metering()
 
-		gasUsed := metering.GasUsedByContract()
-		account, _ := context.GetOutputAccount(runtime.GetSCAddress())
-		account.GasUsed = math.AddUint64(account.GasUsed, gasUsed)
+	remainedFromForwarded := uint64(0)
+	account, _ := context.GetOutputAccount(runtime.GetSCAddress())
+	if context.outputState.ReturnCode == vmcommon.Ok {
+		account.GasUsed, remainedFromForwarded = metering.GasUsedByContract()
 		context.outputState.GasRemaining = metering.GasLeft()
+
+		// backward compatibility
+		if !context.host.IsArwenV2Enabled() && account.GasUsed > metering.GetGasProvided() {
+			return context.CreateVMOutputInCaseOfError(arwen.ErrNotEnoughGas)
+		}
+	} else {
+		account.GasUsed = math.AddUint64(account.GasUsed, metering.GetGasProvided())
 	}
 
 	context.removeNonUpdatedCode(context.outputState)
 
+	err := context.checkGas(remainedFromForwarded)
+	if err != nil {
+		return context.CreateVMOutputInCaseOfError(err)
+	}
+
 	return context.outputState
 }
 
+func (context *outputContext) checkGas(remainedFromForwarded uint64) error {
+	if context.host.IsArwenV2Enabled() == false {
+		return nil
+	}
+
+	gasUsed := uint64(0)
+	gasLockSentForward := false
+	for _, outputAccount := range context.outputState.OutputAccounts {
+		gasUsed = math.AddUint64(gasUsed, outputAccount.GasUsed)
```

Comment:
You can pull the computing of `gasUsed` into a separate function (it can return the `gasLockSentForward` also)

Reply:
I would leave it here. gasCheck is a simple function. it follows the similar pattern in elrond-go

Reply:
👍